feat(API): on exception, select most specific error handler available

csojinb opened the pull request:
# Summary of Changes

Rather than selecting error handlers in LIFO order of registration, select the error handler corresponding to the nearest direct ancestor of the exception type raised. So, for example, if an app only adds a custom error handler for the Python ``Exception`` class, and an ``HTTPForbidden`` error is raised, then we use the default handler for ``HTTPError`` rather than the more general ``Exception`` handler (which is the pre-existing behavior). 

This is implemented by storing error handlers on the API object as a dict rather than a list and looking them up using the method resolution order attribute (`__mro__`) on the raised exception class. 

### NOTE: 
~~This commit only includes the actual implementation and does not address testing or documentation. I am seeking implementation feedback before completing those additional changes.~~ Tests and documentation have been added. Thanks @vytas7 for implementation notes!

### BREAKING CHANGE: 
Registration of a new error handler for type `E` will no longer override previously-registered error handlers for subclasses of type `E`. Registration order will no longer matter *except* when multiple error handlers are registered for the exact same exception type, in which case the most recently registered error handler overrides the previous ones. 

# Related Issues

Addresses #1514.

~~I believe it would also allow reversion of https://github.com/falconry/falcon/pull/1599.~~ Not true. I misremembered why #1599 was needed.

# Pull Request Checklist

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](https://github.com/falconry/falcon/blob/master/CONTRIBUTING.md) at least once; it will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [x] Added **tests** for changed code.
- [x] Prefixed code comments with GitHub nick and an appropriate prefix.
- [x] Coding style is consistent with the rest of the framework.
- [x] Updated **documentation** for changed code.
    - [x] Added docstrings for any new classes, functions, or modules.
    - [x] Updated docstrings for any modifications to existing code.
    - [x] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [x] Updated all relevant supporting documentation files under `docs/`.
    - [x] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [x] Changes (and possible deprecations) have [towncrier](https://pypi.org/project/towncrier/) news fragments under `docs/_newsfragments/`. (Run `towncrier --draft` to ensure it renders correctly.)

If you have *any* questions to *any* of the points above, just **submit and ask**! This checklist is here to *help* you, not to deter you from contributing!

*PR template inspired by the attrs project.*

csojinb commented on falcon/api.py:
```diff
@@ -856,10 +856,44 @@ def _get_body(self, resp, wsgi_file_wrapper=None):
+    # Helpers that do not require self
+    # ------------------------------------------------------------------------
+
+    # QUESTION(csojinb): Find out if this should go here or in `api_helpers` or
```
I wasn't sure if the stuff in `api_helpers.py` is meant to be "public" or not? I don't see any other static methods here, so perhaps it is intended to collect all non-state-required helpers, even if they aren't meant to be "public". If that is the case, I can move this there.

csojinb commented on falcon/api.py:
```diff
@@ -786,22 +787,21 @@ def _handle_exception(self, req, resp, ex, params):
+            return True
+
+        # NOTE(kgriffs): No error handlers are defined for ex and it
+        # is not one of (HTTPStatus, HTTPError, Exception), since it
```
I realized that this note didn't get updated in #1527. I guess it is not too late to add that change there, but I think doing so will override the approval?

vytas7 replied:
I was the only who approved, I can do it again :slightly_smiling_face: 
Not too late IMHO.

csojinb replied:
Heh. Funny how after months and months of that PR sitting there unchanged, I was too late to squeeze in one more change because I left my laptop alone for 4 hours :joy:

kgriffs replied:
Sorry, I guess I jumped the gun. 😄

csojinb replied:
No worries! I did say that it was ready to go.

vytas7 commented on falcon/api.py:
```diff
@@ -770,6 +764,13 @@ def _http_status_handler(self, req, resp, status, params):
         self._compose_error_response(req, resp, error)
 
+    def _find_error_handler(self, ex):
+        for exc in self._get_exception_lineage(ex):
```
We should IMHO inline `type(ex).__mro__[:-1]` here for performance (maybe with a `PERF` note explaining why we are not reusing `get_exception_lineage`).
By saying for performance, I don't just mean avoiding an unnecessary function call which has become cheaper in CPython 3.7+, but also avoiding this check:
```python
        if not isinstance(ex, BaseException):
            raise TypeError('"exception" must be an exception type instance.')
```

It is important to check the type when adding handlers, but when handling an actual exception, we should presume it is of correct type :slightly_smiling_face: Even if in some unfathomable edge case it is not (old-style exceptions? although they were IIRC barred from Python 3), re-raising a `TypeError` would not be any more helpful.

csojinb replied:
Sounds good. I pulled it out into a separate function for readability, but I think we can handle that fine with a `PERF` note as you suggested

csojinb commented:
Would this change warrant a towncrier entry?

csojinb commented on falcon/api.py:
```diff
@@ -532,6 +533,31 @@ def add_error_handler(self, exception, handler=None):
+
+        QUESTION(csojinb): Do we want the usages of falcon.HTTP* below to include links to the error classes?
+
+        If an instance of ``falcon.HTTPForbidden`` is raised, it will be
```
Open questions on the style in this section:

* Should I use single or double backticks for the handler function names? I couldn't decide whether they counted as "variables" or "inline code".
* Should the ussages of `falcon.HTTP*` in this example include `:class:` so that there's a link to the relevant part of the api doc? I didn't do that to start because the specific error classes used for the examples are fairly arbitrary.

csojinb commented on falcon/api.py:
```diff
@@ -532,18 +533,27 @@ def add_error_handler(self, exception, handler=None):
+            app.add_error_handler(Exception, custom_handle_uncaught_exception)
+            app.add_error_handler(falcon.HTTPNotFound, custom_handle_404)
+
+        If an instance of ``falcon.HTTPForbidden`` is raised, it will be
```
Open questions on the style in this section:

* Should I use single or double backticks for the handler function names? I couldn't decide whether they counted as "variables" or "inline code".
 * Should the ussages of falcon.HTTP* in this example include :class: so that there's a link to the relevant part of the api doc? I didn't do that to start because the specific error classes used for the examples are fairly arbitrary.

kgriffs replied:
double backticks are fine, although I tend to add parens to the function name myself to help make it obvious what it is. 

Re class names, I'm fine either way TBH, since you are referencing example code here. But it couldn't hurt to add the `:class:` directive.

vytas7 commented:
@csojinb Re Towncrier I would think, yes, we need an entry (or actually two?) since it is both a breaking change, and a new feature.

mikeyusko commented on falcon/api.py:
```diff
@@ -269,7 +269,8 @@ def __call__(self, env, start_response):  # noqa: C901
                     # HTTP exception signalling the problem, e.g. a 404.
-                    responder, params, resource, req.uri_template = self._get_responder(req)
+                    responder, params, resource, req.uri_template = self._get_responder(
+                        req)
```
Not a big deal, but do we need additional space for `self._get_responder`?

csojinb replied:
You mean, move it to another line because it's past 80 characters? Or, are you asking if this change shouldn't have occurred?

I didn't intentionally make this change here, I think it must have been my auto-pep8. It looks like Falcon's pep8-checker rules allow 99-character lines, so probably I should revert this newline change. I guess my tool didn't pick up the config.

mikeyusko replied:
Yeah, falcon allows 99-characters lines, [flake8 config](https://github.com/falconry/falcon/blob/master/tox.ini#L129).

In that case, I think would be better to revert that newline change

mikeyusko commented on falcon/api.py:
```diff
@@ -879,7 +898,8 @@ def _get_body(self, resp, wsgi_file_wrapper=None):
                 else:
-                    iterable = helpers.CloseableStreamIterator(stream, self._STREAM_BLOCK_SIZE)
+                    iterable = helpers.CloseableStreamIterator(
+                        stream, self._STREAM_BLOCK_SIZE)
```
And the same here:)

csojinb commented:
Yes, I can try to do it this weekend

On Thu, Dec 12, 2019 at 13:44 Kurt Griffiths <notifications@github.com>
wrote:

> *@kgriffs* requested changes on this pull request.
>
> This looks good, do you think you can finish this up soon? I'd love to get
> this merged for the alpha release. AFAICT the remaining TODO items include:
>
>    - Add towncrier news fragments
>    - Rebase on master (fix conflicts)
>    - Tweak docstrings (optional)
>    - Tweak line wrapping (optional)
>
> ------------------------------
>
> In falcon/api.py
> <https://github.com/falconry/falcon/pull/1603#discussion_r357306845>:
>
> >              app.add_error_handler(falcon.HTTPError, custom_handle_http_error)
> -            app.add_error_handler(CustomException)
> +            app.add_error_handler(Exception, custom_handle_uncaught_exception)
> +            app.add_error_handler(falcon.HTTPNotFound, custom_handle_404)
> +
> +        If an instance of ``falcon.HTTPForbidden`` is raised, it will be
>
> double backticks are fine, although I tend to add parens to the function
> name myself to help make it obvious what it is.
>
> Re class names, I'm fine either way TBH, since you are referencing example
> code here. But it couldn't hurt to add the :class: directive.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/falconry/falcon/pull/1603?email_source=notifications&email_token=ABRVDSYBXSEUQM545YHV3I3QYKA7HA5CNFSM4JIGEACKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPAWL7Q#pullrequestreview-331441662>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRVDS5R4LKUAQNI3GFPSEDQYKA7HANCNFSM4JIGEACA>
> .
>

csojinb commented:
Okay, I rebased, undid the overzealous pep8-ing, tweaked the `add_error_handler` docs, and added a towncrier fragment. I think that's everything. cc @kgriffs @vytas7

csojinb commented:
Oh, so it is. I guess kdiff made that decision for me and I didn't look closely enough. I'll fix it and re-push

csojinb commented:
@vytas7 fixed

csojinb commented:
Moving my copied TODO list out of the PR description into this comment:

### TODOs:
- [x] Add towncrier news fragments
- [x] Rebase on master (fix conflicts)
- [x] Tweak docstrings (optional)
- [x] Tweak line wrapping (optional)

(Turns out you can't just check off checkboxes in someone else's comment.)

kgriffs commented:
🎉
